feat: add moxy carousel rfc (#5)
moxystudio/rfcs-oss: rfcs/0005-moxy-carousel.md

- Start Date: 2020-01-21

> - Project: [moxy-carousel](https://github.com/guilhermedeandrade/moxy-carousel)

Reviewer: ```suggestion
- Project: [react-carousel](https://github.com/moxystudio/react-carousel)
```

- Reference Issues: [OFICINA-43](https://moxy.atlassian.net/browse/OFICINA-43)
- Implementation PR: <https://gitlab.com/moxystudio/inoveretail/inovretail-www/merge_requests/102>

> # Summary

Reviewer 2: Should the `Carousel` component perform lazy load? If we have several carousels on the same page it can get heavy

@acostalima

Reviewer: I think it should! @acostalima had already talked with me about this, actually. My bad, I forgot to add this to the discussion. Thanks for pointing this out @ruipneves 👍

This proposal is to use [react-multi-carousel](https://github.com/YIZHUANG/react-multi-carousel) as a baseline to develop our own `Carousel` component. Comparing to other npm packages, namely [nuka-carousel](https://github.com/FormidableLabs/nuka-carousel), [react-slick](https://github.com/akiran/react-slick) and [material-auto-rotating-carousel](https://github.com/TeamWertarbyte/material-auto-rotating-carousel), it's the one that better fits our needs, which are:

> - Infinite mode

Reviewer 3: Include lazy loading in this list.

Reviewer: Added!

- Center mode
- Control mode (arrows/dots)
- Custom animation
- AutoPlay mode
- Responsive
- Swipe to slide
- Mouse drag to slide
- Keyboard control to slide
- Custom styling
- Accessibility support
- Framework agnostic - *this is the only one not offered by `react-multi-carrousel`*

# Basic example

This example was made using [next-with-moxy](https://github.com/moxystudio/next-with-moxy/) boilerplate.

Preview:

> ![moxy-carousel](https://user-images.githubusercontent.com/8797405/72811982-9dc5dc80-3c58-11ea-9a26-aaed7a2e25c4.gif)

Reviewer: ```suggestion
![react-carousel](https://user-images.githubusercontent.com/8797405/72811982-9dc5dc80-3c58-11ea-9a26-aaed7a2e25c4.gif)
```

- `Home.js`

```jsx
import React from 'react';

> import Carousel from 'moxy-carousel';

Reviewer: ```suggestion
import Carousel from '@moxy/react-carousel';
```

import classNames from 'classnames';

import styles from './Home.module.css';

const responsive = {
    desktop: {
        breakpoint: { max: 3000, min: 1024 },

>         items: 3,

Reviewer 2: Does this `items` refer to the number of slide we want to have in each responsive mode?

Reviewer: Exactly!

Reviewer 2: Hmm, so the first 3, 2 ,1 slides will be used? Shouldn't we allow the user to customize this? Maybe I want the pairs only to be shown in mobile (just brainstorming here)

Reviewer: Exactly! Maybe `items`'s value should be an object like
```
const isEven = n => n % 2 === 0

{
 qty: 3,
 filter: isEven
}
```

What you think?

Reviewer: Here we go: https://github.com/moxystudio/react-carousel/pull/5

My proposal to what you suggested! @ruipneves

        slidesToSlide: 1, // Optional, default to 1.

>         partialVisibilityGutter: 40,

Reviewer 2: What is this supposed to be? 😛

Reviewer: When using the `partialVisible` prop, you can define the amount (in pixels) that should be visible with `partialVisibilityGutter`

More info: https://github.com/YIZHUANG/react-multi-carousel#partialvisible-props

    },
    tablet: {
        breakpoint: { max: 1024, min: 464 },
        items: 2,
        slidesToSlide: 2, // Optional, default to 1.
    },
    mobile: {
        breakpoint: { max: 464, min: 0 },
        items: 1,
        slidesToSlide: 1, // Optional, default to 1.
    },
};

const Slide = ({ className, title }) => (
    <div className={ classNames(styles.slide, className) }>{title}</div>
);

const slidesData = [
    { title: 'item 1', className: styles.firstSlide },
    { title: 'item 2', className: styles.secondSlide },
    { title: 'item 3', className: styles.thirdSlide },
    { title: 'item 4', className: styles.forthSlide },
];

const Home = () => (
    <Carousel
        swipeable={ false }
        draggable={ false }

>         showDots

Reviewer 3: We should think of a way to be able to provide a custom component for navigation elements, i.e., dots, arrows, etc. No?

Reviewer: Those are already supported:
- [Custom dots](https://github.com/YIZHUANG/react-multi-carousel#custom-dots)
- [Custom arrows](https://github.com/YIZHUANG/react-multi-carousel#custom-arrows) 
- [Custom button group](https://github.com/YIZHUANG/react-multi-carousel#custom-button-group)

        responsive={ responsive }
        ssr // Means to render carousel on server-side.
        infinite
        autoplay={ false }

>         autoPlaySpeed={ 1000 }

Reviewer 3: It doesn't make much sense to set `autoPlaySpeed` when `autoplay` is `false`...
Should we get rid of `autoplay` and use just keep `autoPlaySpeed` around instead? If `autoPlaySpeed` is `undefined`, the carousel would not autoplay. I think we should also rename the prop to `autoPlaySpeedMs` or similar to indicate the time is provided in milliseconds.

Reviewer: Agreed! I will take a look into the source code to check how feasible this is

        keyBoardControl

>         customTransition="all .5"

Reviewer 2: How is the `customTransition` defined?

Reviewer: As it is right now it can only be defined using inline style. However I think that it shouldn't be too hard to make it work by receiving a class name instead of an inline style.

Reviewer 2: Yes, I think we could control all the animations by making something as `CSSTransition` does. We would send a classname and we could have the `--enter, --active, --leave, --exited` or similar. This way the user can decide the type of transition and define different ones for entering and leaving

            align-items: center;

            color: whitesmoke;
            font-weight: 700;
            text-transform: uppercase;
        }

        & .firstSlide {
            background-color: hotpink;
        }

        & .secondSlide {
            background-color: purple;
        }

        & .thirdSlide {
            background-color: darkcyan;
        }

        & .forthSlide {
            background-color: darksalmon;
        }
    }

    & .activeItemClass {
        opacity: 1;
    }
}
```

# Motivation

Everytime we need to implement a carousel/slider we are dependant on external libs, usually these libs meet the basic requirements, but eventually they will become problematic for more extended use cases.

It would be of great value to have our own implemented carousel/slider to prevent extended overhead time when implementing the wrapper for the external module.

# Detailed design

The original `react-multi-carousel` was lacking a way to style the active item(s) without overriding the package's CSS, so I added a new prop called `activeItemClass` that accepts a `class`.

More changes and/or additions will come after discussion about our needs.

# Drawbacks

> - Depending on how many new features we will add, this will drastically diverge from upstream and thus it will require more effort to maintain;

Reviewer 4: We should make a PR to the react-multi-carousel package with this feature nonetheless. If they don't accept, we keep using this fork.

Reviewer: Sure, will do it!

Reviewer: Done! https://github.com/YIZHUANG/react-multi-carousel/pull/125

- The original codebase is written in TypeScript (although I don't think this is technically a drawback, it should be noted);
- Not compatible with React Native;
- There are 13 contributors, but it seems that the author has been doing almost all of the work.

# Alternatives

- [nuka-carousel](https://github.com/FormidableLabs/nuka-carousel)
- [react-slick](https://github.com/akiran/react-slick)
- [material-auto-rotating-carousel](https://github.com/TeamWertarbyte/material-auto-rotating-carousel)

# Adoption strategy

If we implement this proposal we will need to have in mind that it will require some time to maintain it.

# Unresolved questions

> I'm not sure about [removing the original `react-multi-carousel-item--active`](https://github.com/moxystudio/react-carousel/commit/4f7b87429775f18e455779de38d3b8d64647a688) as the default `class` for an active item. Maybe I should keep it and simply append the `class` given through the `activeItemClass`:

Reviewer 4: Doing that is better. More changes that the PR will be accepted by the repo's owner.

Reviewer: Done! https://github.com/moxystudio/react-carousel/pull/3

> I'm not sure about [removing the original `react-multi-carousel-item--active`](https://github.com/guilhermedeandrade/moxy-carousel/commit/4f7b87429775f18e455779de38d3b8d64647a688) as the default `class` for an active item. Maybe I should keep it and simply append the `class` given through the `activeItemClass`:

Reviewer: ```suggestion
I'm not sure about [removing the original `react-multi-carousel-item--active`](https://github.com/moxystudio/react-carousel/commit/4f7b87429775f18e455779de38d3b8d64647a688) as the default `class` for an active item. Maybe I should keep it and simply append the `class` given through the `activeItemClass`:
```
